parcels/collection/collectionsoa.py

```diff
@@ -26,6 +26,61 @@
     except:
         KMeans = None
 
+#=============================================================================
+# The function partitionParticles4MPI_default() is the default scheme
+# to partition the different particles into different MPI
+# processes. The function setPartitionFunction() will, if called
+# before the particle set is created, allow the user to specify their
+# own scheme for partitioning the particles to different MPI jobs by
+# passing a new partitioning function to setPartitionFunction(). The
+# arguements of this new function must match those of
+# partitionParticles4MPI_default(), as described in the comments to
+# that function.
+
+def partitionParticles4MPI_default(coords,mpi_size=1):
+    '''This function takes the coordinates of the particle starting
+    positions and returns which MPI process will process each
+    particle.
+
+    Input:
+
+    coords: numpy array with rows of [lon, lat] so that
+    coords.shape[0] is the number of particles and coords.shape[1] is 2.
+
+    mpi_size=1: the number of MPI processes.
+
+    Output:
+
+    mpiProcs: an integer array with values from 0 to mpi_size-1
+    specifying which MPI job will run which particles. len(mpiProcs)
+    must equal coords.shape[0]
+
+    '''
+
+    if KMeans:
+        kmeans = KMeans(n_clusters=mpi_size, random_state=0).fit(coords)
+        mpiProcs = kmeans.labels_
+    else:  # assigning random labels if no KMeans (see https://github.com/OceanParcels/parcels/issues/1261)
+        logger.warning_once('sklearn needs to be available if MPI is installed. '
+                            'See http://oceanparcels.org/#parallel_install for more information')
+        mpiProcs = np.randint(0, mpi_size, size=len(lon))
+
+    #print('Using default KMeans partitioning of particles to MPI processes',flush=True)
+
+    return mpiProcs
+
+#by default, the partition is done by the functiond defined above
+partitionParticles4MPI=partitionParticles4MPI_default
+
+#This function, if called before the particle set is created, will alter how the
+#particle set is partitioned between MPI jobs.
+def setPartitionFunction(partitionFunction):
+    global partitionParticles4MPI
+    partitionParticles4MPI=partitionFunction
+    return None
```

Comment:
Would it not make more sense to add a `partitionfunction` argument to the ParticleSet creation? That is more in line with how we treat other options in parcels

e.g. via a 
```python
partition function = kwargs.pop('partition function', partitionParticles4MPI_default)
```

Reply:
I think I have done this successfully. The partition function is now specified with 
```python
pset = pset_type[pset_mode]['pset'].from_list(fieldset, pclass=MyParticle, lon=lonStarts,lat=latStarts,
                                 depth=depthStarts,repeatdt=repeatdt,time=0,
                                 partitionFunction=partitionParticles4MPI_KMeans_constrained)
```